src/e0b30_len_b80.c

```diff
@@ -67,15 +62,139 @@ void bgm_reset_volume(void) {
     gMusicUnkVolume2 = 8;
 }
 
+// Loop struct shenanigans
+#ifdef NON_EQUIVALENT
+
+// Experiment to fix loop
+typedef struct MusicSettingsInner {
+    /* 0x02 */ s16 unk_02;
+    /* 0x04 */ s32 fadeOutTime;
+    /* 0x08 */ s32 fadeInTime;
+    /* 0x0C */ s16 unk_0C;
+    /* 0x0E */ s16 unk_0E;
+    /* 0x10 */ s32 songID;
+    /* 0x14 */ s32 variation;
+    /* 0x18 */ s32 songName;
+    /* 0x1C */ s32 unk_1C;
+    /* 0x20 */ s32 unk_20;
+    /* 0x24 */ s32 unk_24;
+    /* 0x28 */ s32 unk_28;
+    /* 0x2C */ s32 unk_2C;
+} MusicSettingsInner;
+
+typedef struct MusicSettings2 {
+    /* 0x00 */ u16 flags;
+    /* 0x02 */ MusicSettingsInner unk_02;
+} MusicSettings2; // size = 0x30
+
+void bgm_update_music_settings(void) {
+    MusicSettingsInner* inner;
+    MusicSettings2* phi_s1;
+    s32* flags;
+    s32 phi_v0;
+    s32 i;
+    s32 a;
+    s32 c;
+
+    phi_s1 = gMusicSettings;
+
+    for (i = 0; i < 2; i++, phi_s1++) {
+        inner = &phi_s1->unk_02;
+        c = 2;
+        a = ~0x4;
+
+        switch (inner->unk_02) {
+            case 0:
+                break;
+            case 1:
+                if (phi_s1->flags & 1) {
+                    if (inner->fadeOutTime < 250) {
+                        if (!(phi_s1->flags & 4)) {
+                            phi_v0 = func_800559C4(inner->songName);
+                        } else {
+                            phi_v0 = func_80055AF0(inner->songName);
+                        }
+                    } else if (!(phi_s1->flags & 4)) {
+                        phi_v0 = snd_set_song_variation_fade_time(inner->songName, inner->fadeOutTime, 0);
+                    } else {
+                        phi_v0 = func_80055BB8(inner->songName, 0xFA);
+                    }
+                    if (phi_v0 == 0) {
+                        inner->unk_02 = c;
+                    }
+                } else {
+                    if (phi_s1->flags & 4) {
+                        phi_s1->flags |= 0x10;
+                    }
+                    phi_s1->flags &= a;
+                    inner->unk_02 = 5;
+                }
+                break;
+            case 2:
+                phi_s1->flags &= a;
+                if (!(phi_s1->flags & 1)) {
+                    inner->unk_02 = 5;
+                    break;
+                }
+
+                if (func_800559FC(inner->songName) == 0) {
+                    phi_s1->flags &= ~0x1;
+                    inner->unk_02 = 3;
+                }
+
+                break;
+            case 3:
+                inner->unk_02 = 4;
+                break;
+            case 4:
+                inner->unk_02 = 5;
+                break;
+            case 5:
+                if (!(phi_s1->flags & 8)) {
+                    if (inner->songID < 0) {
+                        inner->unk_02 = 0;
+                    } else {
+                        inner->songName = snd_load_song(inner->songID, i);
+                        if (inner->songName > 0xFFFFU) {
```

Comment:
FFFF should be an ASCII_TO_U32 i think

Reply:
what would this be? it's going beyond ascii (0x7E or something) for each char